File: dsa_with_python/heap/max_item_with_given_sum.py
import priority_queue as pq


def max_item_using_sorted_array(arr, sum):
    count = 0
    for i in sorted(arr):
        if i > sum:
            return count
        count += 1
        sum -= i
    return count


def max_item_using_heap(arr, sum):
    for i in arr:
        pq.add(i)
    count = 0
    while not pq.is_empty() and pq.peek() <= sum:
        sum -= pq.poll()
        count += 1
    return count


if __name__ == "__main__":
    arr = [1, 12, 5, 111, 200]
    print(max_item_using_sorted_array(arr, 11))
    print(max_item_using_heap(arr, 11))

File: DsaWithCpp/Heap/MaxItemWithGivenSum.cpp
#include <iostream>
#include <bits/stdc++.h>
#include "PriorityQueue.cpp"
using namespace std;

int maxItemUsingSortedArray(int arr[], int n, int sum)
{
    int count = 0;
    sort(arr, arr + n);
    for (int i = 0; i < n; i++)
    {
        if (arr[i] > sum)
            return count;
        count++;
        sum -= arr[i];
    }
    return count;
}

int maxItemUsingHeap(int arr[], int n, int sum)
{
    PriorityQueue pq;
    for (int i = 0; i < n; i++)
        pq.push(arr[i]);
    int count = 0;
    while (!pq.isEmpty() && pq.top() <= sum)
    {
        sum -= pq.pop();
        count++;
    }
    return count;
}

// int main()
// {
//     int arr[] = {1, 12, 5, 111, 200};
//     cout << maxItemUsingSortedArray(arr, 5, 11) << endl;
//     cout << maxItemUsingHeap(arr, 5, 11) << endl;
// }
